src/system/SystemObject.h

```diff
@@ -204,9 +197,75 @@ class ObjectPool
 public:
     void Reset();
 
-    T * TryCreate();
     void GetStatistics(chip::System::Stats::count_t & aNumInUse, chip::System::Stats::count_t & aHighWatermark);
 
+    template <typename... Args>
+    T * TryCreate(Args &&... args)
+    {
+        T * lReturn = nullptr;
+        static_assert(std::is_base_of<Object, T>::value, "Try to create an object which is not derived from class Object");
+
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+        T * newNode = new T(std::forward<Args>(args)...);
+
+        if (newNode->TryCreate())
+        {
+            std::lock_guard<std::mutex> lock(mMutex);
+            Object * p = &mDummyHead;
+            if (p->mNext)
+            {
+                p->mNext->mPrev = newNode;
+            }
+            newNode->mNext     = p->mNext;
+            p->mNext           = newNode;
+            newNode->mPrev     = p;
+            newNode->mMutexRef = &mMutex;
+            lReturn            = newNode;
+        }
+        else
+        {
+            delete newNode;
+        }
+#else // CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+        unsigned int lIndex = 0;
+
+        for (lIndex = 0; lIndex < N; ++lIndex)
+        {
+            T & lObject = reinterpret_cast<T *>(mArena.uMemory)[lIndex];
+
+            if (lObject.TryCreate())
+            {
+                char buffer[sizeof(T)];
+                new (buffer) T(std::forward<Args>(args)...);
+                size_t octets = sizeof(Object);
+
+                memcpy(reinterpret_cast<char *>(&lObject) + octets, buffer + octets, sizeof(T) - octets);
```

Comment:
`memcpy` an object is not safe, it is undefined behavior. I think you should use placement new using the object address directly.

Reply:
@kghost 

BitMapObjectPool was introduced after SystemPool and its APIs are not thread safe, so BitMapObjectPool could not be used in system layer and inet layer.

BitMapObjectPool always need to be over-provisioned for the worst case, can not reserve and release memory based on its actual usage.

No, you can not use palcement new against the object directly since it will make the API not thread safe, you won't even pass the mutil-threading unit test.

We lock the position in the pool with following atomic operation, after that any operation to this object is thread safe as long as you don't touch the memory reserved by parent class Object.

if (!__sync_bool_compare_and_swap(&this->mRefCount, 0, 1))

You are right, we should not have multiple pool implementation in our system. The goal is to consolidate them into one pool implementation, I have landed several PRs to make SystemPool and BitMapObjectPool compatible, and after this PR, we can consolidate them into single implementation.

Reply:
agree, memcpy solution need to be revisited.

Reply:
In fairness, we were previously memsetting objects, so the direct memory manipulation is not new.

Are there any constraints that can make this viable?  Does any combination of triviality / trivial-copyability make this OK?  Do we know what there constraints were previous to this PR and what the new constraints are with this PR?

Reply:
Maybe the tricky part is storing objects involving deep copying, but this logic is only used in object creation, and memcpy actually is invisible to callers, it seems we are ok even in this case.

Reply:
At least  a comment that summarizes conversation here would help, telling readers or future updates of the code "this is bad, but we did it for this reason..." would be good.

Reply:
Added, I agree the direct memory manipulation is not neat, but using intrusive field to test the availability of the object is more memory efficient since it does not use any extra data structure to manage the static pool.

As Micheal mentioned, this method is not new, and has been tested in CHIP  for a while, I would say we currently continue to use this approach at this moment, it is easy to switch to using additional data structure to record the availability of the object in future if we find any problem with this approach.

Reply:
please add a static check for `std::is_trivially_copyable` if you insist of copying the object.

Reply:
Current intrusive ref count solution put the limitation on the object managed by the ObjectPool to only those objects which are  std::is_trivially_copyable.

The ideal solution is remove this limitation and no use of extra data structure to manage the pool. Not sure if it is feasible. @kghost is working on an improvement solution on top of this one. I will hold this PR for now since it has limitation which we don't want

Reply:
A possibly-safe way to do this:

1. Init the `Object` parts in the array as we do now.
2. Placement-new the object into the temp buffer as now.
3. Copy the parts of the `Object` header we care about out of the array object into the temp object.
4. Use a move `operator=` to move the temp object into the array.  If we make the `Object` `operator=` use atomic ops as needed, this seems like it should be safe...  This does seem like it would be some extra overhead for those atomic ops, though.

Reply:
@bzbarsky-apple Could you explain how 'operator=' make it different?  does this method still suffer the trivial-copyability issue we discussed above, right?

Suppose we try to store the following class into the SystemPool which need deep copying.

class LinkedList {
  LinkedList(LinkedList * prev, LinkedList * next) {
    prev->next = this;
    next->prev = this;
    this->next = next;
    this->prev = prev;
  }
}

'The prev->next = this' will end up with pointing the temp obj we transitionally created.  All the magic we did until now is to avoid using extra data structure such as bitmap to manage the pool.
 
if that means we force all object stored in the SystemPool to implement their own 'operator=' to avoid such case ?

Reply:
> Use a move operator= to move the temp object into the array. If we make the Object operator= use atomic ops as needed, this seems like it should be safe... This does seem like it would be some extra overhead for those atomic ops, though.

I _think_ we are saying that this allows us to assignment copy the entire object, where the part of the object that is Object (base) gets assigned in our special way with our special atomics.  And thus, we don't need to only memcpy the extended part of the object.

This would be reliant on assignment being truly different than copy construction.  I'm getting toward the limits of my personal c++ expertise, but I think these are different in the way we need them to be.

Reply:
> Could you explain how 'operator=' make it different?

Because you can implement it in a way that is aware of your special "don't just copy the memory" semantics.  You don't have to use the default impl.

Copy construction could do this too, of course....

Reply:
"operator =" could avoid memcpy, but this needs all objects inherited from Object implement this operation. Some Pool objects has disabled this operation for reasons.